src/Containers/Microsoft.NET.Build.Containers/Tasks/CreateNewImage.cs

```diff
@@ -325,4 +343,109 @@ public void Dispose()
+            destination.LocalRegistry is not null ? GetLocalStorageType(destination.LocalRegistry) : null);
+        return new Telemetry(Log, context);
+    }
+
+    private RegistryType GetRegistryType(Registry r)
+    {
+        if (r.IsMcr) return RegistryType.MCR;
+        if (r.IsGithubPackageRegistry) return RegistryType.GitHub;
+        if (r.IsAmazonECRRegistry) return RegistryType.AWS;
+        if (r.IsAzureContainerRegistry) return RegistryType.Azure;
+        if (r.IsGoogleArtifactRegistry) return RegistryType.Google;
+        if (r.IsDockerHub) return RegistryType.DockerHub;
+        return RegistryType.Other;
+    }
+
+    private LocalStorageType GetLocalStorageType(ILocalRegistry r)
+    {
+        if (r is ArchiveFileRegistry) return LocalStorageType.Tarball;
+        var d = r as DockerCli;
+        System.Diagnostics.Debug.Assert(d != null, "Unknown local registry type");
+        if (d.GetCommand() == DockerCli.DockerCommand) return LocalStorageType.Docker;
+        else return LocalStorageType.Podman;
+    }
+
+    /// <summary>
+    /// Interesting data about the container publish - used to track the usage rates of various sources/targets of the process
+    /// and to help diagnose issues with the container publish overall.
+    /// </summary>
+    /// <param name="RemotePullType">If the base image came from a remote registry, what kind of registry was it?</param>
+    /// <param name="LocalPullType">If the base image came from a local store of some kind, what kind of store was it?</param>
+    /// <param name="RemotePushType">If the new image is being pushed to a remote registry, what kind of registry is it?</param>
+    /// <param name="LocalPushType">If the new image is being stored in a local store of some kind, what kind of store is it?</param>
+    private record class PublishTelemetryContext(RegistryType? RemotePullType, LocalStorageType? LocalPullType, RegistryType? RemotePushType, LocalStorageType? LocalPushType);
+    private enum RegistryType { Azure, AWS, Google, GitHub, DockerHub, MCR, Other }
+    private enum LocalStorageType { Docker, Podman, Tarball }
+
+    private class Telemetry(Microsoft.Build.Utilities.TaskLoggingHelper Log, PublishTelemetryContext context)
+    {
+        private IDictionary<string, string?> ContextProperties() => new Dictionary<string, string?>
+            {
+                { nameof(context.RemotePullType), context.RemotePullType?.ToString() },
+                { nameof(context.LocalPullType), context.LocalPullType?.ToString() },
+                { nameof(context.RemotePushType), context.RemotePushType?.ToString() },
+                { nameof(context.LocalPushType), context.LocalPushType?.ToString() }
+            };
+
+        public void LogPublishSuccess()
+        {
+            Log.LogTelemetry("sdk/container/publish/success", ContextProperties());
+        }
+
+        public void LogUnknownRepository()
```

Comment:
Do we have an error log for incorrect registry Uris? Is that the same as an unknown repository?

Reply:
We don't - that is very hard to detect because registry URIs can be....any URI? If we can make a good heuristic then this would be an easy error to add tracking for without revamping this whole setup.

An 'unknown repository' in this context is the 'name of the container image'. For example, in `mcr.microsoft.com/dotnet/aspnet`, the registry is 'mcr.microsoft.com' and the repository is 'dotnet/aspnet'.